crates/farmer/ab-proof-of-space-gpu/src/shader/find_matches_and_compute_last.rs

```diff
@@ -0,0 +1,271 @@
+mod cpu_tests;
+#[cfg(all(test, not(miri), not(target_arch = "spirv")))]
+mod gpu_tests;
+
+use crate::shader::compute_fn::compute_fn_impl;
+use crate::shader::constants::{
+    MAX_BUCKET_SIZE, NUM_BUCKETS, NUM_MATCH_BUCKETS, NUM_S_BUCKETS, REDUCED_MATCHES_COUNT,
+};
+use crate::shader::find_matches_in_buckets::rmap::Rmap;
+use crate::shader::find_matches_in_buckets::{
+    LeftTargets, Match, SharedScratchSpace, find_matches_in_buckets_impl,
+};
+use crate::shader::types::{Metadata, Position, PositionY};
+use core::mem::MaybeUninit;
+use spirv_std::arch::{atomic_i_increment, workgroup_memory_barrier_with_group_sync};
+use spirv_std::glam::UVec3;
+use spirv_std::memory::{Scope, Semantics};
+use spirv_std::spirv;
+
+// TODO: Same number as hardcoded in `#[spirv(compute(threads(..)))]` below, can be removed once
+//  https://github.com/Rust-GPU/rust-gpu/discussions/287 is resolved
+pub const WORKGROUP_SIZE: u32 = 256;
+const TABLE_NUMBER: u8 = 7;
+const PARENT_TABLE_NUMBER: u8 = 6;
+
+const _: () = {
+    assert!(crate::shader::find_matches_in_buckets::WORKGROUP_SIZE == WORKGROUP_SIZE);
+};
+
+const PROOFS_BUCKET_SIZE_UPPER_BOUND_SECURITY_BITS: u8 = 128;
+const NUM_ELEMENTS_PER_S_BUCKET: usize =
+    proofs_bucket_upper_bound(PROOFS_BUCKET_SIZE_UPPER_BOUND_SECURITY_BITS) as usize;
+
+/// Calculates a probabilistic upper bound on the bucket size for a given `k`.
+///
+/// Buckets are defined by the lower `NUM_S_BUCKETS.ilog2()` bits of the values. This is based on a
+/// Chernoff bound for the Poisson distribution with mean `lambda = 1`, ensuring the probability
+/// that any bucket exceeds the bound is less than `2^{-security_bits}`. The bound is
+/// `lambda + ceil(sqrt(3 * lambda * (NUM_S_BUCKETS.ilog2() + security_bits) * ln(2)))`.
+/// Accounts for the filter to values in `0..NUM_S_BUCKETS-1` by using the expected number of
+/// remaining elements ~`NUM_S_BUCKETS`, distributed uniformly across all `NUM_S_BUCKETS` buckets.
+const fn proofs_bucket_upper_bound(security_bits: u8) -> u64 {
+    // Lambda is the expected number of entries in a bucket:
+    // ~`NUM_S_BUCKETS / NUM_S_BUCKETS = 1`
+    const LAMBDA: u64 = 1;
+    // Approximation of ln(2) as a fraction: `ln(2) ≈ LN2_NUM / LN2_DEN`.
+    // This allows integer-only computation of the square root term involving ln(2).
+    const LN2_NUM: u128 = 693147;
+    const LN2_DEN: u128 = 1000000;
+
+    // `log2(NUM_S_BUCKETS) + security_bits` for the union bound over `NUM_S_BUCKETS` buckets
+    let ks = NUM_S_BUCKETS.ilog2() as u128 + security_bits as u128;
+    // Compute numerator for the expression under the square root:
+    // `3 * lambda * ks * LN2_NUM`
+    let num = 3u128 * LAMBDA as u128 * ks * LN2_NUM;
+    // Denominator for ln(2): `LN2_DEN`
+    let den = LN2_DEN;
+
+    let ceil_div = num.div_ceil(den);
+
+    // Binary search to find the smallest `x` such that `x * x >= ceil_div`,
+    // which computes `ceil(sqrt(num / den))` without floating-point.
```

Comment:
`isqrt` exists and is const stable as of Rust 1.84:
https://doc.rust-lang.org/std/primitive.u64.html#method.isqrt

Rounding up is trivial, if the square is strictly less than `num / den`, add 1.